### PR TITLE
feat: add stable PrivateIdentifier based on estree spec

### LINKED ISSUE #1436: TypeScript 3.8 Syntax Support
3.8 requires new AST nodes and to support the new syntax. 

### Type Import (#1697) (_SUPPORTED_)
```ts
import type { SomeThing } from "./some-module.js";
export type { SomeThing };
```

### `export * as ns` Syntax (#1698) (_SUPPORTED_)
```TS
export * as utilities from "./utilities.js";
```

### ECMAScript Private Fields (_NOT YET SUPPORTED_ - see #3430)

AST: https://github.com/estree/estree/pull/222

**Important Note**
We currently "support" such that no tooling should fail on private fields.

However, the AST that is produced is not considered stable and ***will*** change in future _without a major release_. Any tooling looking to lint private fields should not rely upon the representation (or be prepared to handle a breakage).

**Why don't we have support yet?**

In short: we are awaiting agreement on the AST representation from the ESTree spec.  

For the 3.7 release we jumped the gun and eagerly implemented optional chaining using babel's AST representation, thinking that the ESTree spec, would just go with that. However ESTree is going with a different and incompatible representation entirely. This will cause our project (and users) a lot of pain as we will have to do a _major_, breaking change to the AST to align with what ESLint core will use.

To avoid going through the same pain again with private class fields, we're holding off on official support until ESTree has decided upon a representation.

There are a lot of projects, and a lot of volunteers that collaborate on the ESTree spec, so things take time; please be patient.

```ts
class Person {
    #name: string
    
    constructor(name: string) {
        this.#name = name;
    }

    greet() {
        console.log(`Hello, my name is ${this.#name}!`);
    }
}

let jeremy = new Person("Jeremy Bearimy");

jeremy.#name
```

### Read more:
https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-beta/

### PR DESCRIPTION
**Pull request by armano2**
Add stable PrivateIdentifier ts-estree base on estree spec,

Due to lack of spec this field field was read as is from typescript,

### Changes to produced AST:

`value` field got renamed to `name` without initial `#`

### TODO:

- [x] add unit tests for `naming-convention` rule
- [ ] add unit tests for `no-unsafe-assignment` rule

---------------

fixes #1436
ref #3430

estree spec: https://github.com/estree/estree/blob/master/experimental/class-features.md

### CODE REVIEW

File: packages/typescript-estree/tests/lib/persistentParse.test.ts [hunk 1]
```diff
@@ -141,7 +141,7 @@ function baseTests(
     fs.mkdirSync(path.join(PROJECT_DIR, 'src', 'bat'));
     fs.mkdirSync(path.join(PROJECT_DIR, 'src', 'bat', 'baz'));
 
-    const bazSlashBar = path.join('bat', 'baz', 'bar') as 'bat/baz/bar';
+    const bazSlashBar = 'bat/baz/bar' as const;
```
Comment (armano2) [inline packages/typescript-estree/tests/lib/persistentParse.test.ts:144]: @bradzacher you can't do `path.join` here as is not going to work on windows machines -> this field is used as key of
CONTENTS

**Comment by bradzacher** [comment]
Based on my discussion with them in https://github.com/estree/estree/issues/240 - I'm happy for us to add these  to the AST.
I'd like to wait until we move them from the experimental folder to be 100% certain nothing is going to change (it shouldn't, but better safe than sorry!)
I'll probs raise a PR there on the weekend.

**Comment by tchetwin** [comment]
@bradzacher I've raised such a PR here: https://github.com/estree/estree/pull/241

**Comment by tchetwin** [comment]
This might make sense in its own Issue rather than this PR, but adding here for continuity of the ESTree class-features spec theme:

Can we reconcile the existing [`ClassProperty`](https://github.com/typescript-eslint/typescript-eslint/blob/2be354bb15f9013a2da1b13a0c0836e9ef057e16/packages/types/src/ast-node-types.ts#L17) AST node with the ESTree proposed spec which spells it [`PropertyDefinition`](https://github.com/estree/estree/blob/master/experimental/class-features.md#propertydefinition)?

**Comment by bradzacher** [comment]
Yes but that is a breaking change, so it would be better as a separate change.

**Comment by tchetwin** [comment]
I've created https://github.com/typescript-eslint/typescript-eslint/issues/3068 to track this following the completion of this PR.
I agree that it would be a breaking change and should be addressed separately.

**Comment by bradzacher** [comment]
Closing in favour of #3808

**Pull request closed without merging**